src/ruby/puppetserver-lib/puppet/server/http_client.rb

```diff
@@ -96,7 +90,14 @@ def create_common_request_options(url, headers, options)
 
     # Ensure multiple requests are not made on the same connection
     headers["Connection"] = "close"
-    request_options = RequestOptions.new(build_url(url))
+
+    if url.is_a?(String)
+      url = URI(url)
+    end
+    url = encode_query(url, params)
+
+    # Java will reparse the string into its own URI object
+    request_options = RequestOptions.new(url.to_s)
```

Comment:
I naively expected JRuby to be able to use the MRI URI directly, so that we didn't have two different URL parser implementations... If `url` has an escaped space, then `to_s` will yield:

```
irb(main):024:0> URI('http://www.example.com/a%20b').to_s
=> "http://www.example.com/a%20b"
```

Does that need to be unencoded before passing it to `RequestOptions.new`? Or does `RequestOptions` assume it's being passed a URL encoded string?

Reply:
The problem that I saw is that `RequestOptions` is not a JRuby class, it's a Java class. So it does accept a URI, but the Java version, not the Ruby version. Alternatively, it takes a string and converts it to the Java URI. Doing some researach, it looks the  constructor does not do any escaping, i.e. it expects things to be properly esacped/encoded already.

Reply:
The alternative would seem to be to create the Java version here in the Ruby code, using the Java URI's piecemeal constructor, which will do encoding on the various bits, but that seems worse?

Reply:
As long as the `to_s` contains the URL encoded path/query, then passing it through as a string should be ok I think. For example to handle the case where the request path contains a `?` character and there is also a query string, then calling `URI#to_s` needs to return `http://www.example.com/a%3Fb?c=d` instead of `http://www.example.com/a?b?c=d` so that we don't "lose" the request path `/a?b` with query `c=d`.